### ISSUE TITLE
Implicit Exec Mask Operand Missing From MCInstrDesc of Some Opcodes in AMDGPU backend

### ISSUE THREAD

**Issue by matinraayai** [llvm/llvm-project#89830]
The following Opcodes do not have `AMDGPU::EXEC` mask as an implicit operand in their MC Instr Desc:
`GLOBAL_STORE_DWORD_vi`
`BUFFER_LOAD_DWORD_OFFEN_vi`
`GLOBAL_STORE_DWORD_SADDR_vi`
The issue appears when constructing a `llvm::MachineInstr` from a valid `llvm::MCInst`, in the same manner as llvm-exegesis here:
https://github.com/llvm/llvm-project/blob/859de94536425376244940e190e069a09d797737/llvm/tools/llvm-exegesis/lib/Assembler.cpp#L157-L176
For the MIR to be correct, simply adding the explicit operands should be enough, as the implicit operands are automatically added according to the MCInstrDesc when calling the `llvm::BuildMI` here:
https://github.com/llvm/llvm-project/blob/859de94536425376244940e190e069a09d797737/llvm/tools/llvm-exegesis/lib/Assembler.cpp#L157
This method will call the `llvm::MachineInstr` constructor with `NoImplicit` flag set to `false`:

https://github.com/llvm/llvm-project/blob/859de94536425376244940e190e069a09d797737/llvm/lib/CodeGen/MachineInstr.cpp#L98-L114

However, when printing the constructed MIR, the implicit exec is nowhere to be seen:
```
GLOBAL_STORE_DWORD_vi $vgpr0_vgpr1, $vgpr2, 0, 0
BUFFER_LOAD_DWORD_OFFEN_vi $vgpr3, $sgpr0_sgpr1_sgpr2_sgpr3, 0, 60, 0, 0
BUFFER_STORE_DWORD_OFFEN_vi $vgpr1, $vgpr3, $sgpr0_sgpr1_sgpr2_sgpr3, 0, 84, 0, 0
```
This causes issues when these instructions get verified before running CodeGen passes.

CC: @arsenm @kzhuravl

**Comment by jayfoad** [comment]
> This causes issues when these instructions get verified before running CodeGen passes.

Why are you trying to run CodeGen passes and MachineVerifier on Real instructions like GLOBAL_STORE_DWORD_vi?

In normal CodeGen, all the passes run on Pseudo instructions like GLOBAL_STORE_DWORD, which does have the implicit use of EXEC. These Pseudo instructions are only converted to the corresponding Real instruction for the appropriate subtarget as part of final code emission.

**Comment by matinraayai** [comment]
@jayfoad the main reason is instrumentation. We use MachinePasses to leverage things like the `AsmPrinter` and pseudo instructions to make our job easier.
Now I can, as you said, somehow convert these opcode to their pseudo equivalent; However:
1. I know there's a way to convert pseudo to MC, I don't know if there's a way to go other way around.
2. I want to keep the lowered MC operands. There's no point adding an extra conversion step here.
3. I've encountered other real instructions that have the EXEC mask as an implicit operand. One example is `V_MOV_B32_e32_vi`. This seems inconsistent with only pseudo instructions having the implicit operands modeled.

IMO even if the normal CodeGen won't encounter real opcodes, that doesn't mean MC shouldn't model the implicit operands correctly. This is especially true if someone is using MC to analyze a binary; They shouldn't need to convert the opcode to its pseudo equivalent first just to get the correctly modeled behavior.

**Comment by jayfoad** [comment]
OK. I certainly have no objection to fixing the implicit uses/defs on Real instructions to match the corresponding Pseudo.

**Comment by arsenm** [comment]
Do the real instructions just not copy the Uses and Defs from the parent pseudo?

**Comment by ritter-x2a** [comment]
> Do the real instructions just not copy the Uses and Defs from the parent pseudo?

Currently, that's not the case for many instructions. This PR changes that: https://github.com/llvm/llvm-project/pull/93004

@matinraayai : feel free to try the commit and check if it solves your problem.

**Comment by ritter-x2a** [comment]
Closing this, as the merged PR #93004 should fix the issue.
